lib/uri_checker.rb

```diff
@@ -0,0 +1,213 @@
+
+    def merge!(other)
+      errors.merge!(other.errors) { |key, oldval, newval| oldval | newval }
+      warnings.merge!(other.warnings) { |key, oldval, newval| oldval | newval }
+    end
+
+    def has_errors?
+      !errors.empty?
+    end
+  end
+
+  class ValidUri
+    HTTP_URI_SCHEMES = %w(http https)
+    FILE_URI_SCHEMES = %w(file)
+
+    attr_reader :report, :options
+
+    def initialize(options = {})
+      @report = Report.new
+      @options = options
+    end
+
+    def call(uri)
+      parsed_uri = URI.parse(uri)
+
+      if parsed_uri.scheme.nil?
+        report.warnings[:no_scheme] << "No scheme given."
+      elsif HTTP_URI_SCHEMES.include?(parsed_uri.scheme)
+        report.merge!(HttpChecker.new(parsed_uri, options).call)
+      elsif FILE_URI_SCHEMES.include?(parsed_uri.scheme)
+        report.merge!(FileChecker.new(parsed_uri, options).call)
+      else
+        report.warnings[:unsupported_scheme] << "Unsupported scheme."
+      end
+
+      report
+    rescue URI::InvalidURIError
+      report.errors[:uri_invalid] << "Invalid URI"
+      report
+    end
+  end
+
+  class HttpChecker
+    INVALID_TOP_LEVEL_DOMAINS = %w(xxx adult dating porn sex sexy singles)
+    REDIRECT_LIMIT = 8
+    REDIRECT_WARNING = 2
+    RESPONSE_TIME_LIMIT = 15
+    RESPONSE_TIME_WARNING = 2.5
+
+    attr_reader :uri, :redirect_history, :report
+
+     def initialize(uri, redirect_history: [])
+      @uri = uri
+      @redirect_history = redirect_history
+      @report = Report.new
+    end
+
+    def call
+      check_redirects
+      return report if report.has_errors?
+
+      check_top_level_domain
+      check_credentials
+
+      head_response = check_head_request
+      return report if report.has_errors?
+
+      check_get_request if head_response && head_response.headers["Content-Type"] == "text/html"
+      return report if report.has_errors?
+
+      check_google_safebrowsing
+
+      report
+    end
+
+  private
+
+    def check_redirects
+      report.errors[:too_many_redirects] << "Too many redirects." if redirect_history.length >= REDIRECT_LIMIT
+      report.errors[:cyclic_redirects] << "Has a cyclic redirect." if redirect_history.include?(uri)
+      report.warnings[:multiple_redirects] << "Multiple redirects." if redirect_history.length == REDIRECT_WARNING
+    end
+
+    def check_top_level_domain
+      tld = uri.host.split(".").last
+      if INVALID_TOP_LEVEL_DOMAINS.include?(tld)
+        report.warnings[:risky_tld] << "Potentially suspicious top level domain."
+      end
+    end
+
+    def check_credentials
+      if uri.user.present? || uri.password.present?
+        report.warnings[:credentials_in_uri] << "Credentials in URI"
+      end
+    end
+
+    def check_head_request
+      start_time = Time.now
+      response = make_request(:head)
+      end_time = Time.now
+      response_time = end_time - start_time
+
+      report.warnings[:slow_response] << "Slow response time" if response_time > RESPONSE_TIME_WARNING
```

Comment:
Possibly not an immediate concern but I wonder if hard-wiring the error and warning messages should get refactored? It feels a bit inflexible. What's the value in storing a symbol and string message here?